### ISSUE TITLE
jnp.sum of all-True boolean array returns wrong result under jax.jit on GPU

### ISSUE THREAD

**Issue by ettrickshepherd** [jax-ml/jax#35762]
## Summary

`jnp.sum(bool_array)` returns `n * 255` instead of `n` when the boolean array is an **all-True (splat) closure-captured constant** under `jax.jit` on GPU. Each `true` value is interpreted as 255 (0xFF) instead of 1 during `convert(i1 → i32)`.

The bug triggers when **all three** conditions are met:
1. The boolean array is **all-True** (represented as a splat `dense<true>` constant in HLO)
2. It is **captured in a JIT closure** (embedded as a constant, not passed as an argument)
3. Execution is on **GPU**

## Minimal reproducer

```python
import jax
import jax.numpy as jnp

mask = jnp.ones(32, dtype=jnp.bool_)  # All-True → splat constant in HLO

print("eager:       ", jnp.sum(mask))                              # 32  ✓
print("jit arg:     ", jax.jit(lambda m: jnp.sum(m))(mask))        # 32  ✓
print("jit closure: ", jax.jit(lambda: jnp.sum(mask))())           # 8160 ✗  (= 32 × 255)
```

## Full test matrix

Tested 15 boolean array constructions as JIT closure constants on GPU. Every **all-True splat** fails with ratio 255×. Every **non-splat** or **all-False splat** passes.

| Test | Construction | Expected | Got | Status |
|---|---|---|---|---|
| `ones_32` | `jnp.ones(32, dtype=bool)` | 32 | **8160** | FAIL |
| `ones_64` | `jnp.ones(64, dtype=bool)` | 64 | **16320** | FAIL |
| `ones_256` | `jnp.ones(256, dtype=bool)` | 256 | **65280** | FAIL |
| `full_true_32` | `jnp.full(32, True)` | 32 | **8160** | FAIL |
| `broadcast_true_32` | `jnp.broadcast_to(True, (32,))` | 32 | **8160** | FAIL |
| `array_list_32` | `jnp.array([True]*32)` | 32 | **8160** | FAIL |
| `computed_all_true` | `jnp.arange(32) < 32` | 32 | **8160** | FAIL |
| `int32_cast` | `jnp.sum(mask.astype(jnp.int32))` | 32 | **8160** | FAIL |
| `computed_partial` | `jnp.arange(32) < 16` | 16 | 16 | OK |
| `mixed_31t_1f` | `[True]*31 + [False]` | 31 | 31 | OK |
| `mixed_1f_31t` | `[False] + [True]*31` | 31 | 31 | OK |
| `mixed_32t_224f` | `[True]*32 + [False]*224` | 32 | 32 | OK |
| `zeros_32` | `jnp.zeros(32, dtype=bool)` | 0 | 0 | OK |
| `single_true` | `jnp.array([True])` | 1 | 1 | OK |
| `single_false` | `jnp.array([False])` | 0 | 0 | OK |

`single_true` (1 element) and `zeros_32` (all-False splat) are fine — only all-True splats with >1 element trigger the bug.

Explicit `.astype(jnp.int32)` before sum does not help — the closure-captured mask is still an i1 splat constant, and the same `convert` op is emitted.

## XLA pass isolation

Disabling HLO optimization passes has **no effect**:

| `XLA_FLAGS` | Result | Fixed? |
|---|---|---|
| (default) | 8160 | No |
| `--xla_disable_hlo_passes=constant_folding` | 8160 | No |
| `--xla_disable_hlo_passes=constant_folding,algsimp` | 8160 | No |

This rules out constant folding and algebraic simplification. The bug is in GPU code generation or the runtime kernel.

## HLO

The lowered StableHLO for `jax.jit(lambda: jnp.sum(mask))` is correct:

```mlir
func.func private @_jit_closure_sum() -> tensor<i32> {
    %c = stablehlo.constant dense<true> : tensor<32xi1>
    %0 = stablehlo.convert %c : (tensor<32xi1>) -> tensor<32xi32>
    %c_0 = stablehlo.constant dense<0> : tensor<i32>
    %1 = stablehlo.reduce(%0 init: %c_0) applies stablehlo.add across dimensions = [0]
         : (tensor<32xi32>, tensor<i32>) -> tensor<i32>
    return %1 : tensor<i32>
}
```

`dense<true> : tensor<32xi1>` is a splat constant. The `convert` from `i1` to `i32` should map `true → 1`, but the GPU produces `true → 255` (0xFF).

## Analysis

`8160 = 32 × 255` — every element contributes 255 instead of 1. The consistent 255× multiplier across all array sizes (32, 64, 256) indicates the GPU `convert(i1 → i32)` kernel reads each `i1` element as a full byte (0xFF) rather than a single bit.

**Hypothesis:** When a `DenseElementsAttr` is a splat (single stored value broadcast to all elements), the GPU materializes or interprets the `i1` constant incorrectly. A splat `true` stores a single byte `0xFF` (or `0x01`), and the convert kernel reads 8 bits per element instead of 1 bit, yielding 255 per element.

This doesn't affect:
- **Non-splat constants** — stored with proper per-element bit indexing
- **Function arguments** — transferred as correctly-typed device buffers, not embedded HLO constants
- **CPU backend** — different constant materialization path
- **All-False splats** — `0x00` reads as 0 regardless of byte vs bit interpretation

## Environment

- **JAX**: 0.9.1
- **jaxlib**: 0.9.1
- **GPU**: NVIDIA L4 (CUDA 12)
- **Platform**: Linux x86_64

CPU (macOS, Linux) is not affected — all tests pass.

## Workaround

Pass the boolean array as a JIT **argument** instead of capturing it in the closure:

```python
# WRONG on GPU:
@jax.jit
def f():
    return jnp.sum(bool_mask)

# Correct:
@jax.jit
def f(mask):
    return jnp.sum(mask)
f(bool_mask)
```

Or compute the reduction eagerly and capture the scalar result:

```python
n = int(jnp.sum(bool_mask))  # Eager: correct
@jax.jit
def f():
    return n  # int closure constant, not i1
```

## Reproducer script

<details>
<summary>Standalone script (no dependencies beyond JAX)</summary>

```python
"""Standalone reproducer: i1 splat convert bug on GPU.

Usage: python repro.py
"""
import json, sys
import jax, jax.numpy as jnp, numpy as np

def _test(name, mask):
    n_true = int(np.asarray(mask).sum())
    eager = int(jax.device_get(jnp.sum(mask)))
    jit_arg = int(jax.device_get(jax.jit(lambda m: jnp.sum(m))(mask)))
    jit_closure = int(jax.device_get(jax.jit(lambda: jnp.sum(mask))()))
    ok = jit_closure == n_true
    entry = {"expected": n_true, "eager": eager, "jit_arg": jit_arg,
             "jit_closure": jit_closure, "ok": ok}
    if not ok and n_true > 0:
        entry["ratio"] = jit_closure / n_true
    print(f"  [{'OK' if ok else 'FAIL'}] {name}: expected={n_true} got={jit_closure}"
          + (f" (ratio={entry['ratio']})" if "ratio" in entry else ""))
    return entry

print(f"JAX {jax.__version__} | {jax.default_backend()} | {jax.devices()}")
results = {}
# Splat all-True (should fail on GPU)
results["ones_32"] = _test("jnp.ones(32, bool)", jnp.ones(32, dtype=jnp.bool_))
results["ones_64"] = _test("jnp.ones(64, bool)", jnp.ones(64, dtype=jnp.bool_))
results["ones_256"] = _test("jnp.ones(256, bool)", jnp.ones(256, dtype=jnp.bool_))
results["full_true"] = _test("jnp.full(32, True)", jnp.full(32, True, dtype=jnp.bool_))
results["array_list"] = _test("jnp.array([True]*32)", jnp.array([True]*32))
results["computed"] = _test("arange(32)<32", jnp.arange(32) < 32)
# Non-splat (should pass)
results["mixed"] = _test("31T+1F", jnp.array([True]*31 + [False]))
results["partial"] = _test("arange(32)<16", jnp.arange(32) < 16)
results["zeros"] = _test("zeros(32, bool)", jnp.zeros(32, dtype=jnp.bool_))
results["single"] = _test("array([True])", jnp.array([True]))

failures = [k for k, v in results.items() if not v["ok"]]
print(f"\n{len(results)} tests, {len(failures)} failures: {failures}")
sys.exit(1 if failures else 0)
```

</details>

**Comment by jakevdp** [comment]
Thanks for the report – seems like it may be related to https://github.com/openxla/xla/issues/37159

**Comment by btgaskin** [comment]
That sounds right, and it's good this is already underway. While I’ve worked around it for now, this was a tricky bug to trace in practice.

If the XLA fix may take some time, I can open a narrow JAX PR for this issue as an interim mitigation:

- avoid `i1 -> int` convert in bool reduction preprocessing by materializing 0/1 via `select`
- add a regression test for closure-captured all-True bool under jit on GPU

I’ve validated this on the reproducer and targeted GPU diagnostics (including closure `32 -> 8160` baseline and `32 -> 32` with the mitigation), though I haven’t run the full cross-backend test matrix yet.

Note that this would only cover the bool reduction path and wouldn't address `convert_element_type` more broadly. I can look into a frontend fix for that also if there's interest, though at a glance it seems more complex.

**Comment by jakevdp** [comment]
It looks like this was introduced recently: I can reproduce with `jax[cuda12]==0.9.1` on T4 GPU, but not with `jax[cuda12==0.9.0]`.

**Comment by jakevdp** [comment]
Looking more closely here, there was a similar bug uncovered by our CI around the 0.9.1 release, addressed by this StableHLO commit: https://github.com/openxla/xla/commit/e88aed8b668136e8b1dc39b80b6d5a416cbc1681

However I tried installing nightly jaxlib and the bug still reproduces.

**Comment by jakevdp** [comment]
Bisection identified https://github.com/openxla/xla/commit/1d1e81f4ac113d1adf9e8e21d6732bfcff25c28e as the change that led to this breakage. cc/ @GleasonK

**Comment by GleasonK** [comment]
```mlir
# File from `print(jax.jit(lambda: jnp.sum(mask)).lower()._lowering.stablehlo())`
$ cat /tmp/t.mlir 
module @jit__lambda attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() -> (tensor<i32> {jax.result_info = "result"}) {
    %c = stablehlo.constant dense<true> : tensor<32xi1>
    %0 = stablehlo.convert %c : (tensor<32xi1>) -> tensor<32xi32>
    %c_0 = stablehlo.constant dense<0> : tensor<i32>
    %1 = stablehlo.reduce(%0 init: %c_0) applies stablehlo.add across dimensions = [0] : (tensor<32xi32>, tensor<i32>) -> tensor<i32>
    return %1 : tensor<i32>
  }
}

# Roundtrip through serialization
$ stablehlo-translate /tmp/t.mlir --serialize --target=1.0.0 | stablehlo-translate-g3 --deserialize
module @jit__lambda attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() -> (tensor<i32> {jax.result_info = "result"}) {
    %c = stablehlo.constant dense<true> : tensor<32xi1>
    %0 = stablehlo.convert %c : (tensor<32xi1>) -> tensor<32xi32>
    %c_0 = stablehlo.constant dense<0> : tensor<i32>
    %1 = stablehlo.reduce(%0 init: %c_0) applies stablehlo.add across dimensions = [0] : (tensor<32xi32>, tensor<i32>) -> tensor<i32>
    return %1 : tensor<i32>
  }
}

# Roundtrip and call optimizers
$ stablehlo-translate /tmp/t.mlir --serialize --target=1.14.0 | stablehlo-opt -vhlo-legalize-to-stablehlo --stablehlo-target-independent-optimization
module @jit__lambda attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() -> (tensor<i32> {jax.result_info = "result"}) {
    %c = stablehlo.constant dense<0> : tensor<i32>
    %c_0 = stablehlo.constant dense<255> : tensor<32xi32> //  <- fold of i1 wrong
    %0 = stablehlo.reduce(%c_0 init: %c) applies stablehlo.add across dimensions = [0] : (tensor<32xi32>, tensor<i32>) -> tensor<i32>
    return %0 : tensor<i32>
  }
}
```

Looks like serialization is handing this alright, but the in-memory representation after deser may be off or at least handled wrong in convert constant folder. Looking into fixes now

**Comment by GleasonK** [comment]
Filed upstream bug as well. After bytecoding we represent splats in-memory as 0xFF (before they're 0x01) so conversion via sign extension is giving the wrong result. Note: This may impact mosaic as well given that its both a StableHLO & upstream issue.

**Comment by jakevdp** [comment]
Confirmed fixed as of v0.10.2 (and probably earlier)
